File: AIOUSB/samples/USB-AI16-16/scripting/sample.pl
#!/usr/bin/perl

use AIOUSB;

{
    package Device;

    sub new(@) {
        my ($self,%args) = @_;
        bless \%args,Device;
    }

}


my $deviceIndex = 0;
my $MAX_NAME_SIZE = 20;

my $deviceIndex = 0;
my $deviceFound = $AIOUSB::AIOUSB_FALSE;

my $CAL_CHANNEL = 5;
my $MAX_CHANNELS = 128;
my $NUM_CHANNELS = 16;
my $NUM_OVERSAMPLES = 10;
my $NUM_SCANS = 100000;
my $BULK_BYTES = $NUM_SCANS * $NUM_CHANNELS * 2 * ($NUM_OVERSAMPLES+1);
my $CLOCK_SPEED = 500000 / ( $NUM_CHANNELS * ($NUM_OVERSAMPLES+1) );


AIOUSB::AIOUSB_Init();
AIOUSB::AIOUSB_Reset( $deviceIndex );
AIOUSB::AIOUSB_SetCommTimeout( $deviceIndex, 1000 );
AIOUSB::AIOUSB_SetDiscardFirstSample( $deviceIndex, $AIOUSB::AIOUSB_TRUE );

$deviceIndex = 0;
$result = AIOUSB::AIOUSB_Init();
if ( $result != AIOUSB::AIOUSB_SUCCESS ) {
    print STDERR "Error running AIOUSB_Init()...";
    exit(1);
}

$deviceMask = AIOUSB::GetDevices();
if( $deviceMask == 0 ) {
    print STDERR "No ACCES devices found on USB bus\n";
    exit(1);
}

$devices = [];
$number_devices = 1;
AIOUSB::AIOUSB_ListDevices();
$index = 0;

while( $deviceMask > 0 && $#{$devices} < $number_devices ) { 
    if( $deviceMask & 1 ) {
        my $obj = AIOUSB::AIODeviceInfoGet( $index );

        if ( ($obj->swig_PID_get() >= $AIOUSB::USB_AI16_16A && $obj->swig_PID_get() <= $AIOUSB::USB_AI12_128E) ||
             ($obj->swig_PID_get() >= $AIOUSB::USB_AIO16_16A &&  $obj->swig_PID_get() <= $AIOUSB::USB_AIO12_128E )) {
            push( @{$devices}, new Device( "index" => $index, "productID"=>$obj->swig_PID_get(), "numDIOBytes"=>$obj->swig_DIOBytes_get,"numCounters" => $obj->swig_Counters_get ));
            print "";
        }
    }
    $index += 1;
    $deviceMask >>= 1;
}

print "Size of devices is $#{$devices}\n";


$deviceIndex = $devices->[0]->{index};


AIOUSB::AIOUSB_Reset( $deviceIndex );
print "Setting timeout\n";
AIOUSB::AIOUSB_SetCommTimeout( $deviceIndex, 1000 );
AIOUSB::AIOUSB_SetDiscardFirstSample( $deviceIndex, $AIOUSB::AIOUSB_TRUE );

$result = 0;
($ndevice,$result) = AIOUSB::AIODeviceTableGetDeviceAtIndex( $deviceIndex );
$cb = AIOUSB::AIOUSBDeviceGetADCConfigBlock( $ndevice );
print "";


AIOUSB::AIOUSB_SetAllGainCodeAndDiffMode( $cb, $AIOUSB::AD_GAIN_CODE_10V, $AIOUSB::AIOUSB_FALSE );
AIOUSB::ADCConfigBlockSetCalMode( $cb, $AIOUSB::AD_CAL_MODE_NORMAL );
AIOUSB::ADCConfigBlockSetTriggerMode( $cb, 0 );
AIOUSB::ADCConfigBlockSetScanRange( $cb, 2, 13 );
AIOUSB::ADCConfigBlockSetOversample( $cb, 0 );



AIOUSB::AIOUSBDeviceWriteADCConfig( $ndevice, $cb );

print "A/D settings successfully configured\n";



$retval = AIOUSB::ADC_SetCal($deviceIndex, ":AUTO:");
if( $result != $AIOUSB::AIOUSB_SUCCESS ) {
    print "Error '%s' performing automatic A/D calibration" % ( AIOUSB_GetResultCodeAsString( result ) );
    sys.exit(0);
}

AIOUSB::ADC_SetOversample( $deviceIndex, 0 );
AIOUSB::ADC_SetScanLimits( $deviceIndex, $CAL_CHANNEL, $CAL_CHANNEL );
AIOUSB::ADC_ADMode( $deviceIndex, 0 , $AIOUSB::AD_CAL_MODE_GROUND );

$counts = AIOUSB::ushortarray->new( 16 );
$result = AIOUSB::ADC_GetScan( $deviceIndex, $counts->cast() );

if( $result < $AIOUSB::AIOUSB_SUCCESS ) {
    print sprintf "Error '%s' attempting to read ground counts\n" , AIOUSB::AIOUSB_GetResultCodeAsString( $result );
} else {
    print sprintf "Ground counts = %u (should be approx. 0)\n" , $counts->{$CAL_CHANNEL} ;
}


AIOUSB::ADC_ADMode( $deviceIndex, 0 , $AIOUSB::AD_CAL_MODE_REFERENCE ); # TriggerMode
$result = AIOUSB::ADC_GetScan( $deviceIndex, $counts->cast() );
if( result < AIOUSB_SUCCESS ) { 
    print sprintf "Error '%s' attempting to read reference counts\n" , AIOUSB_GetResultCodeAsString( result );
} else {
    print sprintf "Reference counts = %u (should be approx. 65130)\n", $counts->{$CAL_CHANNEL};
}

$gainCodes = [ map { 0 } 1..16 ];

#
# demonstrate scanning channels and measuring voltages
#

for($i = 0; $i < 16; $i ++ ) { 
    $gainCodes->[$i] = $AIOUSB::AD_GAIN_CODE_0_10V;
}

AIOUSB::ADC_RangeAll( $deviceIndex , $gainCodes, $AIOUSB::AIOUSB_TRUE );

AIOUSB::ADC_SetOversample( $deviceIndex, $NUM_OVERSAMPLES );
AIOUSB::ADC_SetScanLimits( $deviceIndex, 0, $NUM_CHANNELS - 1 );
AIOUSB::ADC_ADMode( $deviceIndex, 0 , $AD_CAL_MODE_NORMAL );



print "Volts read:\n";

if ( 1 ) { 
# $volts = [map { 0 } 1..16 ];
$volts = AIOUSB::doublearray->new(16);
for( $i = 0; $i < 1 ; $i ++ ) {
    $result = AIOUSB::ADC_GetScanV( $deviceIndex, $volts->cast() );
    for( $j = 0; $j < 16 ; $j ++ ) {
        print sprintf "  Channel %2d = %6.6f\n" , $j, $volts->{$j};
    }
}


}

# demonstrate reading a single channel in volts
$voltage = AIOUSB::new_udp();
$result = AIOUSB::ADC_GetChannelV( $deviceIndex, $CAL_CHANNEL, $voltage );

print sprintf("Result from A/D channel %d was %f\n" , $CAL_CHANNEL, AIOUSB::udp_value($voltage) );


$result = AIOUSB::ADC_GetChannelV( $deviceIndex, 1, $voltage );

print sprintf("Result from A/D channel %d was %f\n", 1, AIOUSB::udp_value($voltage) );


AIOUSB::AIOUSB_Reset( $deviceIndex );
AIOUSB::AIOUSB_Exit();



__END__

File: AIOUSB/samples/USB-AI16-16/scripting/sample.py
#!/usr/bin/env python

import sys
import time
import math
import os
sys.path.append(os.environ["AIO_LIB_DIR"] + "/wrappers/python/build/lib.linux-x86_64-2.7")


import AIOUSB
from AIOUSB import *

class Device:
    name = ""
    serialNumber = 0
    index = 0
    numDIOBytes = 0
    numCounters = 0
    productID = 0
    def __init__(self, **kwds):
        self.__dict__.update(kwds)


MAX_NAME_SIZE = 20;

deviceIndex = 0;
deviceFound = AIOUSB_FALSE

CAL_CHANNEL = 5
MAX_CHANNELS = 128
NUM_CHANNELS = 16
NUM_OVERSAMPLES = 10
NUM_SCANS = 100000
BULK_BYTES = NUM_SCANS * NUM_CHANNELS * 2 * (NUM_OVERSAMPLES+1);
CLOCK_SPEED = 500000 / ( NUM_CHANNELS * (NUM_OVERSAMPLES+1) );

print """USB-AI16-16A sample program
AIOUSB library version %s, %s
This program demonstrates controlling a USB-AI16-16A device on
the USB bus. For simplicity, it uses the first such device found
on the bus.
""" % ( AIOUSB_GetVersion(), AIOUSB_GetVersionDate() )

def find_ai(obj):
    if (obj.PID >= USB_AI16_16A and obj.PID <= USB_AI12_128E) or \
       (obj.PID >=  USB_AIO16_16A and obj.PID <= USB_AIO12_128E ):
        return True

# Simplified API call that uses Python callback that takes 
# an AIOUSBDevice as it's argument .
# Typically, all you care about is the obj.PID matching 

options = NewAIOCommandLineOptionsFromDefaultOptions( AIO_SCRIPTING_OPTIONS() )


device_indices = []
#AIOUSB_FindDevicesByGroup( device_indices, AIOUSB.AIO_ANALOG_INPUT() )
AIOUSB_FindDeviceIndicesByGroup( device_indices, AIOUSB.AIO_ANALOG_INPUT() ) 

# or
# device_indices = AIOUSB_FindDevices( find_ai )

AIOProcessCommandLine( options, sys.argv )

if options.index == -1:
    deviceIndex = device_indices[0]
elif set(device_indices) & set([options.index]):
    deviceIndex = (set(device_indices) & set([options.index])).pop()
else:
    sys.stderr.write("Error: can't find a valid device (index=%d) != (found device indices=%s)\n" % ( options.index,device_indices ))
    sys.exit(1)
                    

AIOUSB_Reset( deviceIndex );
print "Setting timeout"
AIOUSB_SetCommTimeout( deviceIndex, 1000 );

AIOUSB_SetDiscardFirstSample( deviceIndex, AIOUSB_TRUE );

serialNumber = 0

[ndevice,result] = AIODeviceTableGetDeviceAtIndex( deviceIndex )

if result < AIOUSB_SUCCESS:
    print "Error: Can't get device\n"
    exit(1)


cb = AIOUSBDeviceGetADCConfigBlock( ndevice )

AIOUSB_SetAllGainCodeAndDiffMode( cb, AD_GAIN_CODE_10V, AIOUSB_FALSE );
ADCConfigBlockSetCalMode( cb, AD_CAL_MODE_NORMAL );
ADCConfigBlockSetTriggerMode( cb, 0 );
ADCConfigBlockSetScanRange( cb, 2, 13 );
ADCConfigBlockSetOversample( cb, 0 );


AIOUSBDeviceWriteADCConfig( ndevice, cb ); # Write the config block to the device

print "A/D settings successfully configured"

retval = 0
retval = ADC_SetCal(deviceIndex, ":AUTO:")

if retval < AIOUSB_SUCCESS:
    print "Error '%s' performing automatic A/D calibration" % ( AIOUSB_GetResultCodeAsString( retval ) )
    sys.exit(0)

ADCConfigBlockSetScanRange( cb, CAL_CHANNEL, CAL_CHANNEL )
ADCConfigBlockSetTriggerMode( cb , 0 )
ADCConfigBlockSetCalMode( cb , AD_CAL_MODE_GROUND )
AIOUSBDeviceWriteADCConfig( ndevice, cb )

# A better API is coming soon, so you won't have to do
# this to get Data
counts = new_ushortarray( 16 )
result = ADC_GetScan( deviceIndex, counts.cast() );

if retval < AIOUSB_SUCCESS:
    print "Error '%s' attempting to read ground counts\n" % ( AIOUSB_GetResultCodeAsString( result ) )
else:
    print "Ground counts = %u (should be approx. 0)" % ( counts[CAL_CHANNEL] )


ADC_ADMode( deviceIndex, 0 , AD_CAL_MODE_REFERENCE ) # TriggerMode
result = ADC_GetScan( deviceIndex, counts.cast() );
if result < AIOUSB_SUCCESS:
    print "Error '%s' attempting to read reference counts" % ( AIOUSB_GetResultCodeAsString( result ) )
else:
    print "Reference counts = %u (should be approx. 65130)" % ( counts[CAL_CHANNEL] )

gainCodes = [0 for x in range(0,16)]

# 
# demonstrate scanning channels and measuring voltages
# 
for channel in range(0,len(gainCodes)):
    gainCodes[channel] = AD_GAIN_CODE_0_10V


ADC_RangeAll( deviceIndex, gainCodes, AIOUSB_TRUE );


ADC_SetOversample( deviceIndex, NUM_OVERSAMPLES );
ADC_SetScanLimits( deviceIndex, 0, NUM_CHANNELS - 1 );
ADC_ADMode( deviceIndex, 0 , AD_CAL_MODE_NORMAL );

# volts = [0.0 for x in range(0,16)]
volts = doublearray(16)
for i in range(0,1):
    result = ADC_GetScanV( deviceIndex, volts.cast() );
    for j in range(0,16):
        print "  Channel %2d = %6.6f" % ( j, volts[j] )


voltage = new_udp()
result = ADC_GetChannelV( deviceIndex, CAL_CHANNEL, voltage );
print "Result from A/D channel %d was %s " % (CAL_CHANNEL, udp_value(voltage) )





